core/src/main/scala/akka/persistence/cassandra/CassandraStatements.scala

```diff
@@ -28,3 +29,38 @@ private[akka] class CassandraStatements(val settings: PluginSettings) {
   }
 
 }
+
+/**
+ * INTERNAL API: caching to avoid repeated check via ColumnDefinitions
+ */
+@InternalApi private[akka] class ColumnDefinitionCache {
+  private def hasColumn(column: String, row: Row, cached: Option[Boolean], updateCache: Boolean => Unit): Boolean = {
+    cached match {
+      case Some(b) => b
+      case None =>
+        val b = row.getColumnDefinitions.contains(column)
+        updateCache(b)
+        b
+    }
+  }
+
+  @volatile private var _hasMetaColumns: Option[Boolean] = None
+  private val updateMetaColumnsCache: Boolean => Unit = b => _hasMetaColumns = Some(b)
+  def hasMetaColumns(row: Row): Boolean =
+    hasColumn("meta", row, _hasMetaColumns, updateMetaColumnsCache)
+
+  @volatile private var _hasOldTagsColumns: Option[Boolean] = None
+  private val updateOldTagsColumnsCache: Boolean => Unit = b => _hasOldTagsColumns = Some(b)
+  def hasOldTagsColumns(row: Row): Boolean =
+    hasColumn("tag1", row, _hasOldTagsColumns, updateOldTagsColumnsCache)
+
+  @volatile private var _hasTagsColumn: Option[Boolean] = None
+  private val updateTagsColumnCache: Boolean => Unit = b => _hasTagsColumn = Some(b)
+  def hasTagsColumn(row: Row): Boolean =
+    hasColumn("tags", row, _hasTagsColumn, updateTagsColumnCache)
```

Comment:
Above already existed in EventDeserializer. Moved to separate class and added below.